File: src/main/java/catmoe/fallencrystal/moefilter/network/handler/PlayerHandler.java
package catmoe.fallencrystal.moefilter.network.handler;

import catmoe.fallencrystal.moefilter.network.pipeline.IPipeline;
import catmoe.fallencrystal.moefilter.network.util.ChannelExceptionCatcher;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPipeline;
import lombok.Getter;
import net.md_5.bungee.BungeeCord;
import net.md_5.bungee.ConnectionThrottle;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.api.config.ListenerInfo;
import net.md_5.bungee.chat.ComponentSerializer;
import net.md_5.bungee.connection.InitialHandler;
import net.md_5.bungee.netty.ChannelWrapper;
import net.md_5.bungee.protocol.PacketWrapper;
import net.md_5.bungee.protocol.packet.EncryptionResponse;
import net.md_5.bungee.protocol.packet.Handshake;
import net.md_5.bungee.protocol.packet.Kick;
import net.md_5.bungee.protocol.packet.LoginRequest;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.util.HashMap;
import java.util.Map;

public class PlayerHandler extends InitialHandler implements IPipeline {
    public PlayerHandler(final ChannelHandlerContext ctx, final ListenerInfo listener, final ConnectionThrottle throttler) {
        super(BungeeCord.getInstance(), listener);
        this.ctx = ctx;
        this.throttler = throttler;
    }

    private final ConnectionThrottle throttler;
    public InetAddress inetAddress;
    public final ChannelHandlerContext ctx;
    public ChannelPipeline pipeline;


    @Override
    public void connected(final ChannelWrapper wrapper) throws Exception { super.connected(wrapper); pipeline = wrapper.getHandle().pipeline(); inetAddress = ((InetSocketAddress) getSocketAddress()).getAddress(); }

    @Override
    public void exception(final Throwable t) throws Exception {ChannelExceptionCatcher.INSTANCE.handle(ctx.channel(), t);}

    @Override
    public void handle(final PacketWrapper packet) throws Exception {
        if (packet == null) { return; }
        // TODO PacketLimit here if (packet.buf.readableBytes() > ?(Int) {packet.buf.clear(); }
    }

    @Override
    public void handle(final EncryptionResponse response) throws Exception { /* ENCRYPTION PER SEC */ super.handle(response); }

    @Override
    public void handle(final Handshake handshake) throws Exception {
        switch (handshake.getRequestedProtocol()) {
            case 1: { /* Ping */ }
            case 2: { /* Join */ }
            default: { /* Invalid Protocol */ }
        }

        // Blacklist check here
    }

    private static final Map<String, Kick> kickPacketCache = new HashMap<>();
    public void disconnect_(final String reason) {
        if (reason != null && ctx.channel().isActive()) {
            cache: {
                Kick packet = kickPacketCache.get(reason);
                if (packet != null) {ctx.channel().writeAndFlush(packet);break cache;}
                packet = new Kick(ComponentSerializer.toString(new TextComponent(reason)));
                kickPacketCache.put(reason, packet);
                ctx.channel().writeAndFlush(packet);
            }
        }
        ctx.close();
    }

    @Override
    public void handle(final LoginRequest loginRequest) throws Exception {
        if (throttler != null && throttler.throttle(getSocketAddress())) { ctx.close(); return; }
        // TODO Verify here.
    }

    @Override
    public String toString() { return "§7(§f" + getSocketAddress() + (getName() != null ? "|" + getName() : "") + "§7) <-> MoeFilter InitialHandler"; }
}

File: src/main/java/catmoe/fallencrystal/moefilter/network/bungee/handler/PlayerHandler.kt
package catmoe.fallencrystal.moefilter.network.bungee.handler

import catmoe.fallencrystal.moefilter.network.bungee.pipeline.IPipeline
import catmoe.fallencrystal.moefilter.network.bungee.pipeline.IPipeline.Companion.LAST_PACKET_INTERCEPTOR
import catmoe.fallencrystal.moefilter.network.bungee.pipeline.IPipeline.Companion.PACKET_INTERCEPTOR
import catmoe.fallencrystal.moefilter.network.bungee.pipeline.MoeChannelHandler
import catmoe.fallencrystal.moefilter.network.bungee.util.ExceptionCatcher.handle
import catmoe.fallencrystal.moefilter.network.bungee.util.exception.InvalidHandshakeStatusException
import catmoe.fallencrystal.moefilter.network.bungee.util.exception.InvalidStatusPingException
import catmoe.fallencrystal.moefilter.network.bungee.util.exception.PacketOutOfBoundsException
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelPipeline
import net.md_5.bungee.BungeeCord
import net.md_5.bungee.api.config.ListenerInfo
import net.md_5.bungee.connection.InitialHandler
import net.md_5.bungee.netty.ChannelWrapper
import net.md_5.bungee.netty.PipelineUtils
import net.md_5.bungee.protocol.PacketWrapper
import net.md_5.bungee.protocol.packet.*
import java.net.InetAddress
import java.net.InetSocketAddress
import java.util.concurrent.CompletableFuture

class PlayerHandler(
    private val ctx: ChannelHandlerContext,
    listenerInfo: ListenerInfo,
) : InitialHandler(BungeeCord.getInstance(), listenerInfo), IPipeline {
    private var currentState = ConnectionState.HANDSHAKE
    private var inetAddress: InetAddress? = null
    private var pipeline: ChannelPipeline? = null
    @Throws(Exception::class)
    override fun connected(wrapper: ChannelWrapper) {
        super.connected(wrapper)
        pipeline = wrapper.handle.pipeline()
        inetAddress = (socketAddress as InetSocketAddress).address
    }

    @Throws(Exception::class)
    override fun exception(t: Throwable) { handle(ctx.channel(), t) }

    @Throws(Exception::class)
    override fun handle(packet: PacketWrapper?) {
        if (packet == null) { throw PacketOutOfBoundsException("Null packet") }
        if (packet.buf.readableBytes() > 2048) {
            packet.buf.clear()
            throw PacketOutOfBoundsException("Packet" + packet.buf.readableBytes() + "is out of bounds")
        }
    }

    @Throws(Exception::class)
    override fun handle(handshake: Handshake) {
        if (currentState !== ConnectionState.HANDSHAKE) { throw InvalidHandshakeStatusException("") }
        currentState = ConnectionState.PROCESSING
        currentState = when (handshake.requestedProtocol) {
            1 -> { ConnectionState.STATUS }
            2 -> { ConnectionState.JOINING }
            else -> { throw InvalidHandshakeStatusException("Invalid handshake protocol ${handshake.requestedProtocol}") }
        }
        pipeline!!.addBefore(PipelineUtils.BOSS_HANDLER, PACKET_INTERCEPTOR, PacketHandler(this))
        pipeline!!.addLast(LAST_PACKET_INTERCEPTOR, MoeChannelHandler.EXCEPTION_HANDLER)
        try { super.handle(handshake) }
        catch (exception: Exception) { exception.printStackTrace(); ctx.channel().close() }
    }

    private var hasRequestedPing = false
    private var hasSuccessfullyPinged = false
    @Throws(Exception::class)
    override fun handle(statusRequest: StatusRequest) {
        if (hasRequestedPing || hasSuccessfullyPinged || currentState !== ConnectionState.STATUS) { throw InvalidStatusPingException() }
        hasRequestedPing = true
        currentState = ConnectionState.PROCESSING
       CompletableFuture.runAsync {
           if (!isConnected) { throw InvalidStatusPingException() }
           currentState = ConnectionState.PINGING
           hasSuccessfullyPinged = true
           try { super.handle(statusRequest) } catch (_: NoSuchElementException) {} // Actually inject netty failed.
        }
    }

    @Throws(Exception::class)
    override fun handle(ping: PingPacket) {
        if (currentState !== ConnectionState.PINGING || !hasRequestedPing || !hasSuccessfullyPinged) { throw InvalidStatusPingException() }
        currentState = ConnectionState.PROCESSING
        unsafe().sendPacket(ping)
        ctx.close()
    }

    @Throws(Exception::class)
    override fun handle(loginRequest: LoginRequest) {
        if (currentState !== ConnectionState.JOINING) { throw InvalidHandshakeStatusException("") }
        super.handle(loginRequest)
    }

    override fun toString(): String {
        // return "§7(§f" + socketAddress + (if (name != null) "|$name" else "") + "§7) <-> MoeFilter InitialHandler"
        return "§7(§f$socketAddress${if (name != null) "|$name" else ""}§7) <-> MoeFilter InitialHandler"
    }
}
